MayssaraOmar/assignment1/Problem 2.cpp

```diff
@@ -0,0 +1,178 @@
+
+//ASSUMPTIONS
+/*
+- the "kth_to_last" function is a part of the linkedlist class
+- the "kth_to_last" function assumes the list size is unknown
+- the "kth_to_last" function takes O(n) time and 0(1) space.
+*/
+
+#include <iostream>
+using namespace std;
+
+template <class T>
+class linkedlist {
+
+    struct Node {
+        T data;
+        Node *next;
+        Node(const T &input, Node *next) : data(input), next(next) {}
+    };
+    int listsize;
+    Node *head;
+    Node *tail;
+
+
+    struct empty_list_exception : public exception {
```

Comment:
Just a side note to be careful with exceptions. They are a powerful tool, but as most tools in the c++ toolkit it's easy to do more harm than good :)

Reply:
It's the first time I use exceptions in C++. Is this a good exceptions implementation ? And what is the best way to handle these kind of errors instead of using exceptions?

Reply:
I'm not sure, I haven't used them at all here :)
Looks reasonable.
Usually, we use a bool return indicating success and use out parameters for the real return values. Another option is when returning pointers you can use nullptr to indicate a failure.
When things get more complicated, you can return a 'result' type, that carries information about WHAT went wrong in case of error.